ffc/codegeneration/finite_element.py

```diff
@@ -307,34 +307,93 @@ def transform_reference_basis_derivatives(L, ir, parameters):
         msg = "Using %s transform to map values back to the physical element." % mapping.replace(
             "piola", "Piola")
 
+        # List of dof types that require multiplying by -1 if their entity has been reflected
+        # TODO: check that these are all vector and that no other types are vector
+        vector_types = ["PointScaledNormalEval", "ComponentPointEval", "PointEdgeTangent",
+                        "PointFaceTangent", "PointScaledNormalEval", "PointNormalEval"]
+        # For each dof that needs reflection this will contain the entity number that the dof is associated with.
+        # Entities are numbered: Point0, Point1, ..., Edge0, Edge1, ..., Face0, [Face1, ..., Volume]
+        # If no reflection needed, this will be -1
+        entities_of_reflected_dofs = [-1 for i in range(ir.space_dimension)]
+        reflect = False
+        ent_n = 0
+        # Run through the entities and mark vector dofs on each entity
+        for dim, e_dofs in ir.entity_dofs.items():
+            for n, dofs in e_dofs.items():
+                for d in dofs:
+                    if dim > 0:
+                        t = ir.dof_types[d]
+                        if t in vector_types:
+                            entities_of_reflected_dofs[d] = ent_n
+                            reflect = True
+                ent_n += 1
+
         mapped_value = L.Symbol("mapped_value")
-        transform_apply_code += [
-            L.ForRanges(
-                dofrange,
-                (s, 0, num_derivatives_t),
-                (i, 0, num_physical_components),
-                index_type=index_type,
-                body=[
-                    # Unrolled application of mapping to one physical component,
-                    # for affine this automatically reduces to
-                    #   mapped_value = reference_values[..., reference_offset]
-                    L.Comment(msg),
-                    L.VariableDecl(
-                        "const double", mapped_value,
-                        M_scale * sum(
-                            M_row[jj] * reference_values[ip, idof, s, reference_offset + jj]
-                            for jj in range(num_reference_components))),
-                    # Apply derivative transformation, for order=0 this reduces to
-                    # values[ip,idof,0,physical_offset+i] = transform[0,0]*mapped_value
-                    L.Comment("Mapping derivatives back to the physical element"),
-                    L.ForRanges((r, 0, num_derivatives_g),
-                                index_type=index_type,
-                                body=[
-                                    L.AssignAdd(values[ip, idof, r, physical_offset + i],
-                                                transform[r, s] * mapped_value)])
-                ])
-        ]
 
+        # If at least one vector dof needs reflecting
+        if reflect:
```

Comment:
Can you reduce the code duplication here, below? If reflect=True then prepare the conditional statement, otherwise just multiply with literal 1.0 for example.